### ISSUE TITLE
Link SVN Commit User to SCM User

### ISSUE THREAD

**Issue by Shaoranlaos** [scm-manager/scm-manager#1619]
#### Problem to be solved

The user of the SVN commit is not linked to the user profile of the scm user like it is in a git repository.

#### Proposed solution

Link the user of SVN commits via the user name to the users profile page like it is with GIT commits (In the Commit view).

SVN Commit Page:
![image](https://user-images.githubusercontent.com/25054970/115003493-23f5c080-9ea6-11eb-80b4-3e71572b674c.png)

GIT Commit Page:
![image](https://user-images.githubusercontent.com/25054970/115003646-4b4c8d80-9ea6-11eb-84c4-380473527a06.png)

**Comment by pfeuffer** [comment]
I've got your point. The thing is, we do no linking at all. For Git commits we only use the information from the commit itself. In Git (like Mercurial) the author is stored with a username and an email address, unlike in SVN, where there is only a username (therefore you only get an email link in SCM-Manager for Git of HG Commits, no link to a user page or whatsoever).

What would be possible to do (and what out of question would be nice) is a lookup of the username in SVN in the SCM-Manager user database and to use this user's name and email instead of the user id from SVN. Of course this only works, when a matching user is available.

**Comment by sdorra** [comment]
> What would be possible to do (and what out of question would be nice) is a lookup of the username in SVN in the SCM-Manager user database and to use this user's name and email instead of the user id from SVN (what is what you have suggested). Of course this only works, when a matching user is available.

And this is exactly what the [scm-authormapping-plugin](https://scm-manager.org/plugins/scm-authormapping-plugin) does.

**Comment by pfeuffer** [comment]
Thanks for the hint, @sdorra , forgot about this one. Can you give it a try, @Shaoranlaos ? You can install this plugin via the plugin section in the global administration view.

**Comment by Shaoranlaos** [comment]
> For Git commits we only use the information from the commit itself. In Git (like Mercurial) the author is stored with a username and an email address, unlike in SVN, where there is only a username

Right i had forgotten that it only links to the mail for some reason I imagined it linked to the profile of the user.

But the plugin is not what I meant. This would mean I had to administrate every user again.

In SVN you already need a valid username in SCM to make a commit. Couldn't this username be linked to the username in SCM and than automatically mapped to the Full Name so that this can be displayed in the commit list? Or would that need to much performance?

Note: The screenshots above are only examples in a private test instance, the real case would be an extenal user that has a username from an external system (e.g. Active Directory sAMAccount = 8characters).
So it would be more clear who made the commit (for most usernames the brain does that "automatic" but for some usernames that can take some seconds to make sense).

**Comment by sdorra** [comment]
> In SVN you already need a valid username in SCM to make a commit. Couldn't this username be linked to the username in SCM and than automatically mapped to the Full Name so that this can be displayed in the commit list? Or would that need to much performance?

The plugin supports username lookups in the scm-manager database. Todo so you only have to enable the "Auto Mapping" mode in the author mapping plugin configuration.

**Comment by Shaoranlaos** [comment]
I have installed and activated like described and made a new commit but nothing has changed in the shown commit name. It continues to only show the username and not the full name.
Here the Author-Mapping config of the test-repository:
![image](https://user-images.githubusercontent.com/25054970/115185692-a5389780-a0e0-11eb-8799-23368bd53a76.png)
That is the checkbox you mean i hope?

**Comment by sdorra** [comment]
Yes, this is the checkbox i mean. Could you please check that your svn commit author name matches the one in the scm-manager user database (users tab in scm-manager) and that the user has a displayName assigned?

**Comment by Shaoranlaos** [comment]
Yes that is both the case, the user is an external user from Active Directory and i made the commit with this user.

**Comment by Shaoranlaos** [comment]
OK i made a trace log with my commit and see that it tries to map the user but can not find it in the userdatabase.
A bit strang because that is the user i'm logging into SCM with...
`09:16:14.276 [ScmEventBus-1-2] TRACE sonia.scm.authormapping.MappingPrivilegedAction - could not find user with username *****`

**Comment by pfeuffer** [comment]
Do you have the user in the user view in SCM Manager. Do you use the LDAP plugin for AD? This should generate an "external" whenever you log in, and this user should be found by the author mapping.

**Comment by Shaoranlaos** [comment]
Yes the user is in the user view and i use the LDAP Plugin which creates the users when they login the first time.

**Comment by sdorra** [comment]
Sorry, but we can't reproduce the problem. Does the problem occur on every user? Or it is just on special?

**Comment by Shaoranlaos** [comment]
It is only the one user. I tried it with other user and they all work.
I hav compared the users and only 3 things are different (that i found):
- The user is a global admin (in SCM).
- The user has public GPG Keys assigned
- The user is not flagged external but accepts an external (AD) password (conversion doesn't stick it reverts to internal on next login)

Any Idea on the external flag problem? Could this be the cause?

**Comment by Shaoranlaos** [comment]
OK the problem with the external flag fixed itself after a restart. So then i have no idea what could cause the issue.
I also made  another user to a global admin and for this user it is working.

**Comment by sdorra** [comment]
Does the username have special characters in it?

**Comment by Shaoranlaos** [comment]
No (they are Windows usernames so [a-z]{8})
But i found another one where it is not working.
But that one is a standard user without any rights (only WRITE on one repository) and no specialities.
Ironicaly they both start with a "c"

**Comment by sdorra** [comment]
It is possible that you send us the the entry from the `users.xml` of such a user?

**Comment by Shaoranlaos** [comment]
Sorry only if you can give me another way that is not public (there are personal information in it).

But I compared a working and non working user and can not see a difference:
- Both have type ldap
- the existing tags are the same (properties is empty)
- external and active is for both true
- the username differs only on the first letter (c to j)
- displayName and mail have values on both entries

**Comment by pfeuffer** [comment]
Can you install the Script Plugin and run the following script under "Administration" - "Scripts" (and replace the name `scmadmin` in line 2 with one of the usernames in question)? Is the user found?

```groovy
um = injector.getInstance(sonia.scm.user.UserManager.class)
sr = um.search(new sonia.scm.search.SearchRequest("scmadmin", true))
sr.forEach({u -> println(u)})
if (sr.isEmpty()) {
    println "not found"
}
```

**Comment by pfeuffer** [comment]
And while we're at it: Can you run this simple script too and tell us, what charset your server uses?

```groovy
println java.nio.charset.Charset.defaultCharset()
```

**Comment by Shaoranlaos** [comment]
Charset is UTF-8

But the usersearch was interesting:
It outputs the entry of scmadmin for the user "c*****", presumably because the mail of scmadmin contains this username.
But it does not output the real user "c*****".

I will try to change the mail of scmadmin to another mail and see if this helps.

**Comment by Shaoranlaos** [comment]
OK that is it seems to be part of the problem.

I changed the mail and deleted another user that had the username as part of its own username (a separated domain admin account that i used for tests) and now it is working ! :)

Thanks for the help to find the cause. Any plans to change something from this discussion and tests? If not you can close this issue from my side.

**Comment by pfeuffer** [comment]
Good news, we found the issue. Thank you for your patience and your help. We will fix this in one of the upcoming releases.

**Comment by pfeuffer** [comment]
The error has to be fixed in [SearchUtil](https://github.com/scm-manager/scm-manager/blob/5f887d4fa8d11f471de981ec78f5e112d1863fcf/scm-core/src/main/java/sonia/scm/search/SearchUtil.java#L168).

**Comment by pfeuffer** [comment]
Fixed with #1627, released with [2.17.0 of SCM-Manager](https://www.scm-manager.org/download/2.17.0/#changelog)
